src/coprocessor/dag/rpn_expr/function.rs

```diff
@@ -32,6 +49,267 @@ impl<T: RpnFunction + ?Sized> RpnFunction for Box<T> {
+        assert_eq!(payload.args_len(), 2);
+
+        if payload.raw_arg_at(0).is_scalar() {
+            if payload.raw_arg_at(1).is_scalar() {
+                Self::eval_2_args_scalar_scalar(
+                    rows,
+                    f,
+                    context,
+                    payload,
+                    payload.raw_arg_at(0).scalar_value().unwrap(),
+                    payload.raw_arg_at(1).scalar_value().unwrap(),
+                )
+            } else {
+                Self::eval_2_args_scalar_vector(
+                    rows,
+                    f,
+                    context,
+                    payload,
+                    payload.raw_arg_at(0).scalar_value().unwrap(),
+                    payload.raw_arg_at(1).vector_value().unwrap(),
+                )
+            }
+        } else {
+            if payload.raw_arg_at(1).is_scalar() {
+                Self::eval_2_args_vector_scalar(
+                    rows,
+                    f,
+                    context,
+                    payload,
+                    payload.raw_arg_at(0).vector_value().unwrap(),
+                    payload.raw_arg_at(1).scalar_value().unwrap(),
+                )
+            } else {
+                Self::eval_2_args_vector_vector(
+                    rows,
+                    f,
+                    context,
+                    payload,
+                    payload.raw_arg_at(0).vector_value().unwrap(),
+                    payload.raw_arg_at(1).vector_value().unwrap(),
+                )
+            }
+        }
+    }
+
+    #[inline(always)]
+    fn eval_2_args_scalar_scalar<Arg0, Arg1, Ret, F>(
+        rows: usize,
+        mut f: F,
+        context: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+        lhs: &ScalarValue,
+        rhs: &ScalarValue,
+    ) -> Result<VectorValue>
+    where
+        Arg0: Evaluable,
+        Arg1: Evaluable,
+        Ret: Evaluable,
+        F: FnMut(&mut EvalContext, RpnFnCallPayload<'_>, &Arg0, &Arg1) -> Result<Ret>,
+    {
+        let mut result = Vec::with_capacity(rows);
+        let lhs = Arg0::borrow_scalar_value(lhs);
+        let rhs = Arg1::borrow_scalar_value(rhs);
+        for _ in 0..rows {
+            result.push(f(context, payload, lhs, rhs)?);
+        }
+        Ok(Ret::into_vector_value(result))
+    }
+
+    #[inline(always)]
+    fn eval_2_args_scalar_vector<Arg0, Arg1, Ret, F>(
+        rows: usize,
+        mut f: F,
+        context: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+        lhs: &ScalarValue,
+        rhs: &VectorValue,
+    ) -> Result<VectorValue>
+    where
+        Arg0: Evaluable,
+        Arg1: Evaluable,
+        Ret: Evaluable,
+        F: FnMut(&mut EvalContext, RpnFnCallPayload<'_>, &Arg0, &Arg1) -> Result<Ret>,
+    {
+        assert_eq!(rows, rhs.len());
+        let mut result = Vec::with_capacity(rows);
+        let lhs = Arg0::borrow_scalar_value(lhs);
+        let rhs = Arg1::borrow_vector_value(rhs);
```

Comment:
How about implementing a trait for `VectorValue`,`ScalarValue` with a function to visit with an index, so we needn't define these 4 independent similar functions.